schema.go

```diff
@@ -0,0 +1,149 @@
+package tf
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+const (
+	NestingModeSingle NestingMode = 0
+	NestingModeList   NestingMode = 1
+	NestingModeSet    NestingMode = 2
+	NestingModeMap    NestingMode = 3
+)
+
+// Schema is used to define the shape of practitioner-provider information,
+// like resources, data sources, and providers. Think of it as a type
+// definition, but for Terraform.
+type Schema struct {
+	// Attributes are the fields inside the resource, provider, or data
+	// source that the schema is defining. The map key should be the name
+	// of the attribute, and the body defines how it behaves. Names must
+	// only contain lowercase letters, numbers, and underscores.
+	Attributes map[string]Attribute
+
+	// Version indicates the current version of the schema. Schemas are
+	// versioned to help with automatic upgrade process. Whenever you have
+	// a change in the schema you'd like to provide a manual migration for,
+	// you should increment that schema's version by one.
```

Comment:
My concerns about putting too much inline:

1. It'll become out of date if best practices change.
2. I've seen provider developers skim the comment blocks because they were given too much info, and miss the important info they were actually looking for. I worry that if we include too much text, we're going to encourage skimming.

Because of that, I'm tempted to say this belongs more in a Learn guide or a best practices document on the website, where we have the space to tease out the nuance and talk about schema versioning in depth.

Reply:
Agree with putting this info in the Godoc. If best practices change, we can update it just as easily, if not more easily, than any external documentation.